server/rest/src/main/java/org/infinispan/rest/Http11To2UpgradeHandler.java

```diff
@@ -27,19 +28,25 @@
    private static final int MAX_INITIAL_LINE_SIZE = 4096;
    private static final int MAX_HEADER_SIZE = 8192;
 
-   private final RestServer restServer;
+   private final int maxContentLength;
+   private final ChannelHandler http11Handler;
+   private final ChannelHandler http20Handler;
+   private final boolean useAlpn;
 
-   Http11To2UpgradeHandler(RestServer restServer) {
+   public Http11To2UpgradeHandler(ChannelHandler http11Handler, ChannelHandler http20Handler, int maxContentLength, boolean useAlpn) {
```

Comment:
We could just `RestServerConfiguration` to set the class variables instead of explicitly passing `maxContentLength` and `useAlpn`?

So:
```java
new Http11To2UpgradeHandler(new Http11RequestHandler(server), new Http20RequestHandler(server), server.getConfiguration().maxContentLength(), server.getConfiguration().ssl().enabled())
```

becomes:
```java
new Http11To2UpgradeHandler(new Http11RequestHandler(server), new Http20RequestHandler(server), server.getConfiguration())
```

Reply:
It seems like a good idea. I'll try it out and see how other bits wire up. If I won't discover any problems, I'll do that!

Reply:
I somewhat reverted the changes I made here. At first I wanted to make REST protocol optional. But this doesn't make any sense since with `HTTP/1.1 Upgrade` and `TLS/ALPN` type of upgrades you should always have a fallback protocol. This effectively means that we should always have a REST Server as a fallback.

Reply:
Could you elaborate on "make REST protocol optional" and "always have a REST Server as a fallback"?

Reply:
It'd be really bad if REST clients such as curl, wget, etc, weren't able to talk to the server since REST is gone :)

Reply:
ooh no, no.. I'm not removing anything ;)

There are two upgrade paths for the client. 

The first one is to "upgrade" existing connection to the new protocol. This method uses `HTTP 1.1/Upgrade` header. This of course requires some "existing connection", so we are assuming you have a client that is already talking to the HTTP Server. This probably requires a REST Server (I could imagine writing a simple HTTP handler which is only capable of upgrading it's connection but using a real REST Server makes more sense to me).

The second method is using `TLS/ALPN`. During TLS handshake the client and the server negotiates connection protocol. There is a change (due to OpenSSL versions mismatch for example) that this negotiation is not properly understood by the server. In such case, you're using a fallback protocol which is `HTTP 1.1`. Again I can imagine using a mock server or something similar, but using a real REST Server makes more sense.

So to sum it up - technically it is possible to use negotiation mechanism without a REST Server. Practically, it doesn't make much sense (at least to me).